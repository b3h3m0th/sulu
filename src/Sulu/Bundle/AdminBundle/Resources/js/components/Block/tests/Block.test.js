// @flow
import React from 'react';
import {render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import log from 'loglevel';
import Block from '../Block';

jest.mock('loglevel', () => ({
    warn: jest.fn(),
}));

jest.mock('../../../utils/Translator', () => ({
    translate: (key) => key,
}));

test('Render an expanded block with multiple types', () => {
    const {container} = render(
        <Block
            activeType="type1"
            expanded={true}
            handle={<span>Test</span>}
            icons={['su-eye', 'su-people']}
            onCollapse={jest.fn()}
            onExpand={jest.fn()}
            onSettingsClick={jest.fn()}
            types={{'type1': 'Type1', 'type2': 'Type2'}}
        >
            Some block content
        </Block>
    );

    expect(container).toMatchSnapshot();
});

test('Render an block without handle or collapse or expand button', () => {
    const {container} = render(
        <Block expanded={true}>
            Some block content
        </Block>
    );
    expect(container).toMatchSnapshot();
});

test('Render a selected block', () => {
    const {container} = render(
        <Block expanded={false} selected={true}>
            Some block content
        </Block>
    );
    expect(container).toMatchSnapshot();
});

test('Render a collapsed block', () => {
    const {container} = render(
        <Block expanded={false} icons={['su-eye', 'su-people']} onCollapse={jest.fn()} onExpand={jest.fn()}>
            Some block content
        </Block>
    );
    expect(container).toMatchSnapshot();
});

test('Do not show type dropdown if only a single type is passed', () => {
    const {container} = render(
        <Block expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()} types={{'type': 'Type'}}>
            Some block content
        </Block>
    );

    // eslint-disable-next-line testing-library/no-container
    const elements = container.getElementsByClassName('select');

    expect(elements).toHaveLength(0);
});

test('Do not show action icon if no actions prop has been passed', () => {
    render(
        <Block expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()} types={{'type': 'Type'}}>
            Some block content
        </Block>
    );

    expect(screen.queryByLabelText('su-more-circle')).not.toBeInTheDocument();
});

test('Do not show action icon if an empty actions prop has been passed', () => {
    render(
        <Block actions={[]} expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()} types={{'type': 'Type'}}>
            Some block content
        </Block>
    );

    expect(screen.queryByLabelText('su-more-circle')).not.toBeInTheDocument();
});

test('Do not show settings icon if no onSettingsClick prop has been passed', () => {
    render(
        <Block expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()} types={{'type': 'Type'}}>
            Some block content
        </Block>
    );

    expect(screen.queryByLabelText('su-cog')).not.toBeInTheDocument();
});

test('Clicking on a collapsed block should call the onExpand callback', async() => {
    const expandSpy = jest.fn();
    render(<Block onCollapse={jest.fn()} onExpand={expandSpy}>Block content</Block>);

    await userEvent.click(screen.queryByRole('switch'));

    expect(expandSpy).toHaveBeenCalledTimes(1);
});

test('Clicking on a expanded block should not call the onExpand callback', async() => {
    const expandSpy = jest.fn();
    render(<Block expanded={true} onCollapse={jest.fn()} onExpand={expandSpy}>Block content</Block>);

    await userEvent.click(screen.queryByRole('switch'));

    expect(expandSpy).not.toBeCalled();
});

test('Clicking the close icon in an expanded block should collapse it', async() => {
    const collapseSpy = jest.fn();
    render(<Block expanded={true} onCollapse={collapseSpy} onExpand={jest.fn()}>Block content</Block>);

    const closeIcon = screen.queryByLabelText('su-collapse-vertical');
    expect(closeIcon).toBeInTheDocument();

    await userEvent.click(closeIcon);

    expect(collapseSpy).toHaveBeenCalledTimes(1);
});

test('Clicking the action icon should open a popover that displays the given actions', async() => {
    const actions = [
        {
            type: 'button',
            icon: 'su-test-1',
            label: 'Test Action 1',
            onClick: jest.fn(),
        },
        {
            type: 'divider',
        },
        {
            type: 'button',
            icon: 'su-test-2',
            label: 'Test Action 2',
            onClick: jest.fn(),
        },
        {
            type: 'button',
            icon: 'su-test-3',
            label: 'Test Action 3',
            onClick: jest.fn(),
        },
    ];
    const {container} = render(
        <Block actions={actions} expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()}>Block content</Block>
    );
    const icon = screen.queryByLabelText('su-more-circle');

    expect(screen.queryByText(/Test Action 1/)).not.toBeInTheDocument();
    expect(icon).toBeInTheDocument();

    await userEvent.click(icon);

    expect(screen.getByText(/Test Action 1/)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});

test('Clicking an action in the action popover should fire the respective callback', async() => {
    const onActionClickSpy = jest.fn();
    const actions = [
        {
            type: 'button',
            icon: 'su-test-1',
            label: 'Test Action 1',
            onClick: onActionClickSpy,
        },
    ];
    render(
        <Block actions={actions} expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()}>Block content</Block>
    );

    const icon = screen.queryByLabelText('su-more-circle');

    await userEvent.click(icon);
    expect(onActionClickSpy).not.toBeCalled();

    await userEvent.click(screen.queryByText('Test Action 1'));
    expect(onActionClickSpy).toBeCalledWith();
});

test('Render remove action if deprecated onRemove prop is set', async() => {
    const removeSpy = jest.fn();
    render(
        <Block expanded={true} onCollapse={jest.fn()} onExpand={jest.fn()} onRemove={removeSpy}>Block content</Block>
    );
    expect(log.warn).toBeCalledWith(
        expect.stringContaining('The "onRemove" prop of the "Block" component is deprecated')
    );

    const actionIcon = screen.queryByLabelText('su-more-circle');
    expect(actionIcon).toBeInTheDocument();
    await userEvent.click(actionIcon);

    const removeIcon = screen.queryByLabelText('su-trash-alt');
    expect(removeIcon).toBeInTheDocument();
    await userEvent.click(removeIcon);

    expect(removeSpy).toHaveBeenCalledTimes(1);
});

test('Changing the type should call the onTypeChange callback', async() => {
    const typeChangeSpy = jest.fn();
    const types = {
        type1: 'Type 1',
        type2: 'Type 2',
    };

    render(
        <Block
            activeType="type1"
            expanded={true}
            onCollapse={jest.fn()}
            onExpand={jest.fn()}
            onTypeChange={typeChangeSpy}
            types={types}
        >
            Block content
        </Block>
    );

    const selectButton = screen.queryByText('Type 1');
    await userEvent.click(selectButton);

    const typeButton = screen.queryByText('Type 2');
    await userEvent.click(typeButton);

    expect(typeChangeSpy).toBeCalledWith('type2');
    expect(typeChangeSpy).toHaveBeenCalledTimes(1);
});
