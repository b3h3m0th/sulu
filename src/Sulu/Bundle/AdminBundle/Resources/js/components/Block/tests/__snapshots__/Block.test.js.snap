// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicking the action icon should open a popover that displays the given actions 1`] = `
<div>
  <section
    class="block expanded"
    role="switch"
  >
    <div
      class="content"
    >
      <header
        class="header"
      >
        <div
          class="iconButtons"
        >
          <button
            type="button"
          >
            <span
              aria-label="su-more-circle"
              class="su-more-circle"
            />
          </button>
          <button
            type="button"
          >
            <span
              aria-label="su-collapse-vertical"
              class="su-collapse-vertical"
            />
          </button>
        </div>
      </header>
      <article
        class="children"
      >
        Block content
      </article>
    </div>
  </section>
</div>
`;

exports[`Render a collapsed block 1`] = `
<div>
  <section
    class="block"
    role="switch"
  >
    <div
      class="content"
    >
      <header
        class="header"
      >
        <div
          class="icons"
        >
          <span
            aria-label="su-eye"
            class="su-eye"
          />
          <span
            aria-label="su-people"
            class="su-people"
          />
        </div>
        <span
          aria-label="su-expand-vertical"
          class="su-expand-vertical"
        />
      </header>
      <article
        class="children"
      >
        Some block content
      </article>
    </div>
  </section>
</div>
`;

exports[`Render a selected block 1`] = `
<div>
  <section
    class="block expanded selected"
    role="switch"
  >
    <div
      class="content"
    >
      <header
        class="header"
      >
        <div
          class="iconButtons"
        />
      </header>
      <article
        class="children"
      >
        Some block content
      </article>
    </div>
  </section>
</div>
`;

exports[`Render an block without handle or collapse or expand button 1`] = `
<div>
  <section
    class="block expanded"
    role="switch"
  >
    <div
      class="content"
    >
      <header
        class="header"
      >
        <div
          class="iconButtons"
        />
      </header>
      <article
        class="children"
      >
        Some block content
      </article>
    </div>
  </section>
</div>
`;

exports[`Render an expanded block with multiple types 1`] = `
<div>
  <section
    class="block expanded"
    role="switch"
  >
    <div
      class="handle"
    >
      <span>
        Test
      </span>
    </div>
    <div
      class="content"
    >
      <header
        class="header"
      >
        <div
          class="types"
        >
          <div
            class="select"
            role="none"
          >
            <button
              class="displayValue default"
              type="button"
            >
              <div
                aria-label="Type1"
                class="croppedText"
                title="Type1"
              >
                <div
                  aria-hidden="true"
                  class="front"
                >
                  Typ
                </div>
                <div
                  aria-hidden="true"
                  class="back"
                >
                  <span>
                    e1
                  </span>
                </div>
                <div
                  class="whole"
                >
                  Type1
                </div>
              </div>
              <span
                aria-label="su-angle-down"
                class="su-angle-down toggle"
              />
            </button>
          </div>
        </div>
        <div
          class="icons"
        >
          <span
            aria-label="su-eye"
            class="su-eye"
          />
          <span
            aria-label="su-people"
            class="su-people"
          />
        </div>
        <div
          class="iconButtons"
        >
          <button
            type="button"
          >
            <span
              aria-label="su-cog"
              class="su-cog"
            />
          </button>
          <button
            type="button"
          >
            <span
              aria-label="su-collapse-vertical"
              class="su-collapse-vertical"
            />
          </button>
        </div>
      </header>
      <article
        class="children"
      >
        Some block content
      </article>
    </div>
  </section>
</div>
`;
